nvm.sh

```diff
@@ -275,6 +275,357 @@ nvm_find_up() {
+              version=substr(comparator,2);
+              split(version, a, /\./);
+              output=output ">=" version " <" a[1]+1 ".0.0 ";
+            }
+          } else if ( match(comparator, /^~/)  ) {
+            if ( match(comparator, /^~0.0.[0-9]+$/) ) {
+              version=substr(comparator,2)
+              split(version, a, /\./)
+              output=output ">=" version " <0." a[2]+1 ".0 "
+            } else if ( match(comparator, /^~0.[0-9]+.[0-9]+$/) ) {
+              version=substr(comparator,2)
+              split(version, a, /\./)
+              output=output ">=" version " <0." a[2]+1 ".0 "
+            } else if ( match(comparator, /^~[0-9]+.[0-9]+.[0-9]+$/) ) {
+              version=substr(comparator,2)
+              split(version, a, /\./)
+              output=output ">=" version " <" a[1] "." a[2]+1 ".0 "
+            }
+          } else {
+            output=output comparator " "
+          }
+        }
+        print output
+      }' |
+
+      # remove leading/trailing spaces
+      command sed -E 's/^ +//;s/ +$//'
+    )
+
+    # only comparator_sets composed of the required grammar are marked as valid
+    if nvm_string_contains_regexp "$validated_comparator_set" '^( ?(<|<=|>|>=)?[0-9]+\.[0-9]+\.[0-9]+)+$'; then
+      validated_semver="$validated_semver || $validated_comparator_set"
+    else
+      return 1
+    fi
+  done
+
+  nvm_echo "$(command printf "%s" "$validated_semver" | command sed -E 's/^ \|\| //')"
+) }
+
+nvm_interpret_complex_semver() { (
+  [ -n "${1-}" ] || return 1
+
+  # Validate incoming semver and transform it into the grammar that is expected by the following logic
+  valid_transformed_semver=$(nvm_validate_semver "${1-}")
+  [ -n "$valid_transformed_semver" ] || ( nvm_err "invalid semantic version: '${1-}'" && return 1 )
+
+  # Iterate through the comparator_sets in the semver.
+  # For each comparator_set, evaluate it to the highest compatible node version.
+  # Output this function with the highest node version among all the node versions collected in the previous step.
+  valid_transformed_semver=$(command printf "%s" "$valid_transformed_semver" | command tr '||' '\n')
+  highest_compatible_node_versions=''
+
+  # TODO add tests verifying this logic correctly gets list of node versions and that the order is from oldest to newest
+  # list of node versions is sorted from oldest to newest
+  remote_node_versions=$(nvm_ls_remote | grep -o 'v[0-9]\+\.[0-9]\+\.[0-9]\+')
+  [ -n "$remote_node_versions" ] || ( nvm_err "failure retrieving remote node versions" && return 1 )
+
+  while [ -n "$valid_transformed_semver" ]; do
+    remote_node_versions_copy=$(command printf "%s" "$remote_node_versions")
+    comparator_set=$(command printf "%s" "$valid_transformed_semver" | head -n1 | command sed -E 's/^ +//;s/ +$//')
+    valid_transformed_semver=$(command printf "%s" "$valid_transformed_semver" | tail -n +2)
+    [ -n "$comparator_set" ] || continue
```

Comment:
please use if statements instead of abusing &&/|| for control flow

Reply:
the `[ -n "$som_variable" ] || continue` pattern for within loops is a preexisting pattern I copied. I am however slightly abusing &&/|| control flow in a few other places though and will changes those.